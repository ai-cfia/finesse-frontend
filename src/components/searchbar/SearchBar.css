.button {
  all: unset;
  cursor: pointer;
}

input {
  background-color: transparent;
  border: none;
  height: 100%;
  width: 100%;
  font-size: 1.2rem;
  margin-left: 10px;
}

input:focus {
  outline: none;
}