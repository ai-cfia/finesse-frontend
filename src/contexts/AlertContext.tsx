import React, {
  createContext,
  useContext,
  useState,
  type ReactNode,
} from "react";

// Type for the alert context state
interface AlertContextState {
  alertMessage: string | null;
  setAlertMessage: React.Dispatch<React.SetStateAction<string | null>>;
}

// Create the context with a default value
const AlertContext = createContext<AlertContextState>({
  alertMessage: null,
  setAlertMessage: () => {},
});

// Create a provider component for the alert context
export const AlertProvider: React.FC<{ children: ReactNode }> = ({
  children,
}) => {
  const [alertMessage, setAlertMessage] = useState<string | null>(null);

  return (
    <AlertContext.Provider value={{ alertMessage, setAlertMessage }}>
      {children}
    </AlertContext.Provider>
  );
};

// Custom hook to use the alert context
export const useAlert = (): AlertContextState => {
  const context = useContext(AlertContext);
  if (context === null) {
    throw new Error("useAlert must be used within an AlertProvider");
  }
  return context;
};
