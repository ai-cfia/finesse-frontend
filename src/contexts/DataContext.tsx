import { createContext, useContext, useState, type ReactNode } from "react";
import { SearchSource, type QueryResult } from "../types";

// Define the type for your context state
interface DataContextState {
  searchTerm: string | null;
  setSearchTerm: React.Dispatch<React.SetStateAction<string | null>>;
  currentSearchSource: SearchSource;
  setCurrentSearchSource: React.Dispatch<React.SetStateAction<SearchSource>>;
  queryResult: QueryResult[];
  setQueryResult: React.Dispatch<React.SetStateAction<QueryResult[]>>;
}

// Create the context with a default value
const DataContext = createContext<DataContextState>({
  searchTerm: null,
  setSearchTerm: () => {},
  currentSearchSource: SearchSource.Simulated,
  setCurrentSearchSource: () => {},
  queryResult: [],
  setQueryResult: () => {},
});

// Create a provider component
export const DataProvider: React.FC<{ children: ReactNode }> = ({
  children,
}) => {
  const [searchTerm, setSearchTerm] = useState<string | null>(null);
  const [currentSearchSource, setCurrentSearchSource] = useState<SearchSource>(
    SearchSource.Simulated,
  );
  const [queryResult, setQueryResult] = useState<QueryResult[]>([]);

  return (
    <DataContext.Provider
      value={{
        searchTerm,
        setSearchTerm,
        currentSearchSource,
        setCurrentSearchSource,
        queryResult,
        setQueryResult,
      }}
    >
      {children}
    </DataContext.Provider>
  );
};

// Custom hook to use the layout context
export const useData = (): DataContextState => {
  const context = useContext(DataContext);
  if (context === null) {
    throw new Error("useDataFetch must be used within a DataFetchProvider");
  }
  return context;
};
