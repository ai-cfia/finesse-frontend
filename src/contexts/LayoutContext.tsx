import React, {
  createContext,
  useContext,
  useEffect,
  useState,
  type ReactNode,
} from "react";

// Type for your context state
interface LayoutContextState {
  isDebugPanelVisible: boolean;
  toggleDebugPanel: () => void;
  isDebugButtonVisible: boolean;
  isAlertBannerVisible: boolean;
  setIsAlertBannerVisible: React.Dispatch<React.SetStateAction<boolean>>;
}

// Create the context with a default value
const LayoutContext = createContext<LayoutContextState>({
  isDebugPanelVisible: false,
  toggleDebugPanel: () => {},
  isDebugButtonVisible: false,
  isAlertBannerVisible: false,
  setIsAlertBannerVisible: () => {},
});

// Create a provider component
export const LayoutProvider: React.FC<{ children: ReactNode }> = ({
  children,
}) => {
  const [isDebugPanelVisible, setIsDebugPanelVisible] = useState(false);
  const [isDebugButtonVisible, setIsDebugButtonVisible] = useState(false);
  const [isAlertBannerVisible, setIsAlertBannerVisible] = useState(false);

  useEffect(() => {
    const debugMode = process.env.REACT_APP_DEBUG_MODE;
    setIsDebugButtonVisible(debugMode?.toLowerCase() === "true");
  }, []);

  const toggleDebugPanel = (): void => {
    console.log("Toggling Debug Panel visibility");
    setIsDebugPanelVisible(!isDebugPanelVisible);
  };

  return (
    <LayoutContext.Provider
      value={{
        isDebugPanelVisible,
        toggleDebugPanel,
        isDebugButtonVisible,
        isAlertBannerVisible,
        setIsAlertBannerVisible,
      }}
    >
      {children}
    </LayoutContext.Provider>
  );
};

// Custom hook to use the layout context
export const useLayout = (): LayoutContextState => {
  const context = useContext(LayoutContext);
  if (context === null) {
    throw new Error("useLayout must be used within a LayoutProvider");
  }
  return context;
};
