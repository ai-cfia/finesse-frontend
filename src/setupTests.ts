// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";

export const setupTestEnvVars = (): void => {
  process.env.REACT_APP_BACKEND_URL = "backend";
  process.env.REACT_APP_BASENAME = "/";
  process.env.REACT_APP_DEBUG_MODE = "True";
  process.env.REACT_APP_GITHUB_API_URL = "github";
};
