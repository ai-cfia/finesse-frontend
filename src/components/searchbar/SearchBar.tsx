import * as React from "react";
import { useEffect, useState } from "react";
import { FaArrowRight, FaSearch } from "react-icons/fa";
import { useNavigate } from "react-router-dom";
import { useStateValue } from "../../StateProvider";
import { ActionTypes, SearchSource } from "../../types";
import "./SearchBar.css";

interface SearchBarProps {
  term?: string;
}

export const SearchBar: React.FC<SearchBarProps> = ({ term }) => {
  const {
    state: { currentSearchSource },
    dispatch,
  } = useStateValue();
  const [searchQuery, setSearchQuery] = useState("");
  const navigate = useNavigate();
  const [previousSearchQuery, setPreviousSearchQuery] = useState<
    string | undefined
  >("");

  useEffect(() => {
    if (typeof term === "undefined") {
      setPreviousSearchQuery("Type to search...");
    } else {
      setPreviousSearchQuery(term);
    }
  }, [term]);

  const search = (e: React.FormEvent): void => {
    e.preventDefault();
    navigate("/search");

    // Ignore dispatch if useSimulatedData is true
    // TODO: revisit
    if (currentSearchSource === SearchSource.Simulated && searchQuery === "") {
      console.log("Search dispatch ignored because useSimulatedData is true");
      return;
    }

    dispatch({
      type: ActionTypes.SetSearchTerm,
      payload: searchQuery,
    });
  };

  return (
    <form className="form-wrapper">
      <div className="input-wrapper">
        <FaSearch id="fa-arrow-right" />
        <input
          placeholder={previousSearchQuery}
          value={searchQuery}
          onChange={(e) => {
            setSearchQuery(e.target.value);
          }}
        />
        <button className="button" type="submit" onClick={search}>
          <FaArrowRight id="fa-arrow-right" style={{ color: "#05486c" }} />
        </button>
      </div>
    </form>
  );
};
