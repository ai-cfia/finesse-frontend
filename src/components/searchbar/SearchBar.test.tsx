import { fireEvent, render, screen } from "@testing-library/react";
import { useNavigate } from "react-router-dom";
import { useData } from "../../contexts/DataContext";
import { setupTestEnvVars } from "../../setupTests";
import { SearchBar } from "./SearchBar";

jest.mock("../../contexts/DataContext", () => ({
  useData: jest.fn(),
}));

jest.mock("react-router-dom", () => ({
  useNavigate: jest.fn(),
}));

describe("SearchBar", () => {
  const mockSetSearchTerm = jest.fn();
  const mockNavigate = jest.fn();

  beforeAll(() => {
    setupTestEnvVars();
  });

  beforeEach(() => {
    (useData as jest.Mock).mockReturnValue({
      searchTerm: "initial",
      setSearchTerm: mockSetSearchTerm,
    });
    (useNavigate as jest.Mock).mockReturnValue(mockNavigate);
    jest.spyOn(console, "warn").mockImplementation(() => {});
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly", () => {
    const { asFragment } = render(<SearchBar />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("updates input value on change", () => {
    const { asFragment } = render(<SearchBar />);
    const input = screen.getByTestId("search-input");
    fireEvent.change(input, { target: { value: "test query" } });
    expect(asFragment()).toMatchSnapshot();
  });

  it("trims search query and navigates on valid form submit", () => {
    const { asFragment } = render(<SearchBar />);
    const input = screen.getByTestId("search-input");
    const button = screen.getByTestId("submit-button");
    fireEvent.change(input, { target: { value: "  test query  " } });
    fireEvent.click(button);
    expect(asFragment()).toMatchSnapshot();
  });

  it("does not navigate or set search term with empty trimmed query", () => {
    const { asFragment } = render(<SearchBar />);
    const input = screen.getByTestId("search-input");
    const button = screen.getByTestId("submit-button");
    fireEvent.change(input, { target: { value: "     " } });
    fireEvent.click(button);
    expect(asFragment()).toMatchSnapshot();
  });

  it("keeps the search query in the input after submission", () => {
    const { asFragment } = render(<SearchBar />);
    const input = screen.getByTestId("search-input");
    const button = screen.getByTestId("submit-button");
    fireEvent.change(input, { target: { value: "test query" } });
    fireEvent.click(button);
    expect(asFragment()).toMatchSnapshot();
  });
});
