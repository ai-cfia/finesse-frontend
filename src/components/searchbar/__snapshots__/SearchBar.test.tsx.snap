// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar does not navigate or set search term with empty trimmed query 1`] = `
<DocumentFragment>
  <form
    class="form-wrapper"
  >
    <div
      class="input-wrapper"
    >
      <svg
        data-testid="search-icon"
        fill="currentColor"
        height="1em"
        id="fa-arrow-right"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
        />
      </svg>
      <input
        data-testid="search-input"
        placeholder="Type to search..."
        value="     "
      />
      <button
        class="button"
        data-testid="submit-button"
        type="submit"
      >
        <svg
          fill="currentColor"
          height="1em"
          id="fa-arrow-right"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(5, 72, 108);"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
          />
        </svg>
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`SearchBar keeps the search query in the input after submission 1`] = `
<DocumentFragment>
  <form
    class="form-wrapper"
  >
    <div
      class="input-wrapper"
    >
      <svg
        data-testid="search-icon"
        fill="currentColor"
        height="1em"
        id="fa-arrow-right"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
        />
      </svg>
      <input
        data-testid="search-input"
        placeholder="Type to search..."
        value="test query"
      />
      <button
        class="button"
        data-testid="submit-button"
        type="submit"
      >
        <svg
          fill="currentColor"
          height="1em"
          id="fa-arrow-right"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(5, 72, 108);"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
          />
        </svg>
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`SearchBar renders correctly 1`] = `
<DocumentFragment>
  <form
    class="form-wrapper"
  >
    <div
      class="input-wrapper"
    >
      <svg
        data-testid="search-icon"
        fill="currentColor"
        height="1em"
        id="fa-arrow-right"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
        />
      </svg>
      <input
        data-testid="search-input"
        placeholder="Type to search..."
        value="initial"
      />
      <button
        class="button"
        data-testid="submit-button"
        type="submit"
      >
        <svg
          fill="currentColor"
          height="1em"
          id="fa-arrow-right"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(5, 72, 108);"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
          />
        </svg>
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`SearchBar trims search query and navigates on valid form submit 1`] = `
<DocumentFragment>
  <form
    class="form-wrapper"
  >
    <div
      class="input-wrapper"
    >
      <svg
        data-testid="search-icon"
        fill="currentColor"
        height="1em"
        id="fa-arrow-right"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
        />
      </svg>
      <input
        data-testid="search-input"
        placeholder="Type to search..."
        value="  test query  "
      />
      <button
        class="button"
        data-testid="submit-button"
        type="submit"
      >
        <svg
          fill="currentColor"
          height="1em"
          id="fa-arrow-right"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(5, 72, 108);"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
          />
        </svg>
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`SearchBar updates input value on change 1`] = `
<DocumentFragment>
  <form
    class="form-wrapper"
  >
    <div
      class="input-wrapper"
    >
      <svg
        data-testid="search-icon"
        fill="currentColor"
        height="1em"
        id="fa-arrow-right"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
        />
      </svg>
      <input
        data-testid="search-input"
        placeholder="Type to search..."
        value="test query"
      />
      <button
        class="button"
        data-testid="submit-button"
        type="submit"
      >
        <svg
          fill="currentColor"
          height="1em"
          id="fa-arrow-right"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(5, 72, 108);"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
          />
        </svg>
      </button>
    </div>
  </form>
</DocumentFragment>
`;
